### DISCUSSION TITLE
Feature request: add SDM72D-M

### DISCUSSION THREAD

**Discussion by TomaszSz7** [arendst/Tasmota#10862] [Ideas]
Hi I would like to rise a request for add the function of reading parameters from SDM72D-M  energy meter.
Maybe someone has already done it?  The official code has only supports for SDM 120 and SDM630.The library is rather the same but the protocal seems to be different.
I would be grateful for the information.
Br! 
Tom

**Comment by barbudor** [comment]
Can you provide a PR ?

**Reply by TomaszSz7** [reply]
Sure. What do you mean PR ? ;-)

**Reply by sfromis** [reply]
Pull Request in Github. Basically it means the relevant working changes to the source code of Tasmota.

**Reply by barbudor** [reply]
a PullRequest means that you have to:
- Fork Tasmota into you own GitHub account
- Download you fork to your PC
- Make the relevant changes and addition, test
- Push back the changes to your own account
- Request Tasmota to pull the new code from your account (the Pull Request)

Let's face that only someone that has the device can write the code and test it.
You have the device, so it's up to you to add the feature in Tasmota.

Now, if the above is overcomplicated for you, but you have all the relevant information, publish all details here and may be someone will pair with you to do the coding and let you do the testing.

**Reply by curzon01** [reply]
I wrote a driver for SDM72D-M it last year, but didn't find it very interesting because this energy meter doesn't seem very common and doesn't provide any other data other than consumption.

If you want I can add it

**Reply by TomaszSz7** [reply]
@curzon01
I basically don't need any other data:-) Would you be so kind and you share the code with me ?

**Comment by curzon01** [comment]
saw I have to adapt it to the current dev so will do a PR for all next days, pls give some time...

**Reply by curzon01** [reply]
@TomaszSz7 - it's in (PR #10892), you can use tasmota-sensor, it's included - give it a try

**Reply by TomaszSz7** [reply]
@curzon01 Thank you very much! I have already tested. Works great!
This is exactly what I was looking for. 
Indeed, this meter doesn't provide much information but I don't need this information either. 
I wanted to thank you once again. You saved my meter. :-)  I already thought it is total c**p. ;-)

**Comment by kugelkopf123** [comment]
Or just Simply use the sml/scripting driver. https://tasmota.github.io/docs/Smart-Meter-Interface/

**Comment by TomaszSz7** [comment]
Hello!
I don't know do you guys know but a new model of SDM72D-M. meter has been released. It has much more functions than the previous one 
SDM72D-M. Link to https://xn--stromzhler-v5a.eu/media/pdf/93/17/d7/SDM72DM-V2.pdf
I accidentally got the second version. ;-)  Question for you guys. Would it be a lot of effort to modify the configuration for SDM630 to work for SDM72D-M v2 ? 

For many values addresses are the same as with SDM630 but after changing to sdm630 ex/tx there is no communication with the meter.. The SDM72 profile also no longer communicates with this meter.
@curzon01  also this time I count on your help . 

Br
Thomas

**Reply by barbudor** [reply]
The registers in the attached PDF are fully matching those of SDM630
Can you try the SDM630 driver ?

**Reply by TomaszSz7** [reply]
@barbudor I've tried.  There is no communication. No infomation on the meter display. The diode in the converter not flash.

**Reply by barbudor** [reply]
Are you sure that your meter is set to modbus adress 1 ?

If the TX led doesn't light, you may have a connection problem
Assuming an ESP8266:
- Pin RX of the adapter should be connected to GPIO 3 and configured as SDM630 RX
- Pin TX of the adapter should be connected to GPIO 1 and configured as SDM630 TX

**Reply by TomaszSz7** [reply]
"Are you sure that your meter is set to modbus adress 1 ?" -> Yes

I using the same esp8266 module which was communicating with SDM72-m v1.
BTW I tried it on ESPHome and it works fine. 
I'm  trying to fix this issue but basically it works fine.
"Phase 1 line to neutral volts': Sending state 0.00000 W with 0 decimals of accuracy"

**Reply by barbudor** [reply]
Can you please set `WebLog 4` and restart
Then copy paste the logs from the boot until 1st Teleperiod message
Thanks

**Reply by TomaszSz7** [reply]
it's quite weird... 
When I added a third cable for ground ( A, B, G ) the connection  has been established. 
Still no value in tasmota
Driver for SDM630

00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F8, Count 20
00:00:00.055 QPC: Count 1
00:00:00.058 CFG: CR 338/699, Busy 0
00:00:00.063 ROT: Mode 1
00:00:00.064 NRG: Init driver 10
00:00:00.066 SRC: Restart
00:00:00.068 Project tasmota - SDM72DM Version 12.2.0(sensors)-2_7_4_9(2022-10-17T08:41:56)
00:00:00.106 SNS: Hardware Serial
00:00:00.181 WIF: Checking connection...
00:00:00.182 WIF: Attempting connection...
00:00:00.518 WIF: Connecting to AP1 ########## Channel 1 BSSId 2C:C8:1B:BC:A1:FB in mode 11n as tasmota-04EABD-2749...
00:00:01.503 WIF: Checking connection...
00:00:01.504 WIF: Connected
00:00:01.754 HTP: Web server active on tasmota-04EABD-2749 with IP address 192.168.0.123
00:00:02.410 WIF: Sending Gratuitous ARP
00:00:02.411 NTP: Sync time...
00:00:02.435 RTC: UTC 2022-10-22T20:18:57, DST 2022-03-27T02:00:00, STD 2022-10-30T03:00:00
21:18:57.000 RTC: Synced by NTP
21:18:57.319 RSL: INFO1 = {"Info1":{"Module":"Generic","Version":"12.2.0(sensors)","FallbackTopic":"cmnd/DVES_04EABD_fb/","GroupTopic":"cmnd/tasmotas/"}}
21:18:57.322 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-04EABD-2749","IPAddress":"192.168.0.123"}}
21:18:57.324 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":8}}
21:18:58.752 DMP: 01 04 04 00 00 00 00 FB 84
21:18:59.003 DMP: 01 04 04 00 00 00 00 FB 84
21:18:59.257 DMP: 01 04 04 00 00 00 00 FB 84
21:18:59.503 DMP: 01 04 04 00 00 00 00 FB 84
21:18:59.815 DMP: 01 04 04 00 00 00 00 FB 84
21:19:00.003 DMP: 01 04 04 00 00 00 00 FB 84
21:19:00.258 DMP: 01 04 04 42 61 2D 19 62 B8
21:19:00.503 DMP: 01 04 04 00 00 00 00 FB 84
21:19:00.753 DMP: 01 04 04 00 00 00 00 FB 84
21:19:01.003 DMP: 01 04 04 00 00 00 00 FB 84
21:19:01.257 DMP: 01 04 04 00 00 00 00 FB 84
21:19:01.464 QPC: Reset
21:19:01.504 DMP: 01 04 04 00 00 00 00 FB 84
21:19:01.753 DMP: 01 04 04 00 00 00 00 FB 84
21:19:02.003 DMP: 01 04 04 00 00 00 00 FB 84
21:19:02.257 DMP: 01 04 04 00 00 00 00 FB 84
21:19:02.415 RSL: STATE = {"Time":"2022-10-22T21:19:02","Uptime":"0T00:00:09","UptimeSec":9,"Heap":22,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"Wifi":{"AP":1,"SSId":"##########","BSSId":"#############","Channel":1,"Mode":"11n","RSSI":64,"Signal":-68,"LinkCount":1,"Downtime":"0T00:00:03"}}
21:19:02.426 RSL: SENSOR = {"Time":"2022-10-22T21:19:02","ENERGY":{"TotalStartTime":"2022-10-22T21:10:44","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":[0,0,0],"Power":[0,0,0],"ApparentPower":[0,0,0],"ReactivePower":[0,0,0],"Factor":[0.00,0.00,0.00],"Voltage":[0,0,0],"Current":[0.000,0.000,0.000]}}
21:19:02.504 DMP: 01 04 04 00 00 00 00 FB 84
21:19:02.530 HTP: Main Menu
21:19:02.753 DMP: 01 04 04 00 00 00 00 FB 84
21:19:03.004 DMP: 01 04 04 00 00 00 00 FB 84
21:19:03.257 DMP: 01 04 04 00 00 00 00 FB 84
21:19:03.411 APP: Boot Count 8
21:19:03.504 DMP: 01 04 04 00 00 00 00 FB 84
21:19:03.753 DMP: 01 04 04 00 00 00 00 FB 84
21:19:04.004 DMP: 01 04 04 00 00 00 00 FB 84
21:19:04.300 CFG: Saved to flash at F7, Count 21, Bytes 4096
21:19:04.303 DMP: 01 04 04 3C 23 D7 0A D9 E9
21:19:04.304 NRG: EnergyTotal[0] 0.0000 kWh
21:19:04.306 NRG: EnergyTotal[1] 0.0000 kWh
21:19:04.307 NRG: EnergyTotal[2] 0.0000 kWh

**Reply by barbudor** [reply]
From what I see in the logs, every values are 0 except
- Phase 1 Active Power
- Total active Energy (which is ignored by Tasmota)

There's maybe a communication problem which makes the meter not returning the proper registers
But the messages received are correct without CRC error otherwise there would be an error log

I have added some logs, can you try this version: [tasmota-sensors-sdm630.zip](https://github.com/arendst/Tasmota/files/9845271/tasmota-sensors-sdm630.zip)
Or if you prefer to compile yourself, here is my branch: https://github.com/barbudor/Tasmota/blob/sdm630/tasmota/tasmota_xnrg_energy/xnrg_10_sdm630.ino

**Reply by curzon01** [reply]
> Hello! I don't know do you guys know but a new model of SDM72D-M. meter has been released.

Yes, I tested this meter too.

I run it well with the new generic modbus energy meter support [xnrg_29_modbus.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xnrg_energy/xnrg_29_modbus.ino) (thx Theo)  enabling by 
```C
#define USE_ENERGY_SENSOR
#define USE_MODBUS_ENERGY
```
and

```haskell
Rule3 on file#modbus do {"Name":"SDM72V2","Baud":9600,"Config":"8N1","Address":1,"Function":4,"Voltage":[0,2,4],"Current":[6,8,10],"Power":[12,14,16],"ApparentPower":[18,20,22],"ReactivePower":[24,26,28],"Factor":[30,32,34],"Frequency":70,"Total":342,"User":[{"R":0x48,"J":"ImportActive","G":"Import Active","U":"kWh","D":24},{"R":0x502,"J":"ExportPower","G":"Export Power","U":"W","D":23},{"R":0x500,"J":"ImportPower","G":"Import Power","U":"W","D":23}]} endon`
```

The only issue I found is that the user defined single phase register are displayed using 3 phase layout output null

**Reply by arendst** [reply]
Great you found it!!!!

I will see how to change the nulls to single column. It's on my list once I finish the shelly pro 2pm.....

**Reply by TomaszSz7** [reply]
@curzon01 Great! 

I have a issue  when I compile the build...
Do you have maybe a compiled file ? 

Compiling .pio\build\tasmota-debug\libb37\ESP8266SdFat\SpiDriver\SdSpiSAM3X.cpp.o
*** [.pio\build\tasmota-debug\src\tasmota.ino.cpp.o] Error 1
======================================================================================================================================== [FAILED] Took 34.05 seconds ========================================================================================================================================

this rule should look like this ? 
on the command line yes ? 

10:06:48.456 CMD: Rule3 1
10:06:48.463 RSL: RESULT = {"Rule3":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":456,"Free":55,"Rules":"on file#modbus do {\"Name\":\"SDM72V2\",\"Baud\":9600,\"Config\":\"8N1\",\"Address\":1,\"Function\":4,\"Voltage\":[0,2,4],\"Current\":[6,8,10],\"Power\":[12,14,16],\"ApparentPower\":[18,20,22],\"ReactivePower\":[24,26,28],\"Factor\":[30,32,34],\"Frequency\":70,\"Total\":342,\"User\":[{\"R\":0x48,\"J\":\"ImportActive\",\"G\":\"Import Active\",\"U\":\"kWh\",\"D\":24},{\"R\":0x502,\"J\":\"ExportPower\",\"G\":\"Export Power\",\"U\":\"W\",\"D\":23},{\"R\":0x500,\"J\":\"ImportPower\",\"G\":\"Import Power\",\"U\":\"W\",\"D\":23}]} endon`"}}

@arendst when you plan to finish shelly pro? :-)

How do you think ,SDM72V2 will be added to the master build ?

**Reply by curzon01** [reply]
> I have a issue when I compile the build... 

Which way? VSC? Clear cache, search for issue using VSC when compiling or simply use the online [compiling tools](https://tasmota.github.io/docs/Compile-your-build/#compiling-tools) way like Gitpod

> this rule should look like this ? on the command line yes ?
> 10:06:48.456 CMD: Rule3 1 

No, do not enable the rule (that's not a normal rule):

- compile tasmota with USE_MODBUS_ENERGY enabled.
- disable the rule `Rule3 0`
- reenter the rule as shown above `Rule3 on file#modbus do {"Name":"SDM72V2","B...`
- reboot device (!)

> How do you think ,SDM72V2 will be added to the master build ?

Why? the generic driver was written for this purpose

**Reply by TomaszSz7** [reply]
> >* compile tasmota with USE_MODBUS_ENERGY enabled.  -> done
> >* disable the rule `Rule3 0` -> done
> >* reenter the rule as shown above `Rule3 on file#modbus do {"Name":"SDM72V2","B...`   - > done
> >* reboot device (!) -> done

console message

19:44:47.040 CMD: Rule3 on file#modbus do {"Name":"SDM72V2","....
19:44:47.067 RUL: Stored uncompressed, would compress from 455 to 351 (-23%)
19:44:47.072 RSL: RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":455,"Free":56,"Rules":"on file#modbus do {\"Name\":\"SDM72V2\",\"Baud\":9600,\"Config\":\"8N1\",\"Address\":1,\"Function\":4,\"Voltage\":[0,2,4],\"Current\":[6,8,10],\"Power\":[12,14,16],\"ApparentPower\":[18,20,22],\"ReactivePower\":[24,26,28],\"Factor\":[30,32,34],\"Frequency\":70,\"Total\":342,\"User\":[{\"R\":0x48,\"J\":\"ImportActive\",\"G\":\"Import Active\",\"U\":\"kWh\",\"D\":24},{\"R\":0x502,\"J\":\"ExportPower\",\"G\":\"Export Power\",\"U\":\"W\",\"D\":23},{\"R\":0x500,\"J\":\"ImportPower\",\"G\":\"Import Power\",\"U\":\"W\",\"D\":23}]} endon"}}


what should happen ? 
on the list of components for GPIO I should see SDM72V2 ?


> >Why? the generic driver was written for this purpose 

ok

**Reply by curzon01** [reply]
> what should happen ? on the list of components for GPIO I should see SDM72V2 ?

No, it's an generic driver so configure your Modbus GPIO using `NrgModbus Tx/Rx`, for example:

![image](https://user-images.githubusercontent.com/6636844/197735141-d814b4bf-a55e-4eb9-b7a6-3063222fd2e5.png)

After that rule3 gets its functionality

**Reply by TomaszSz7** [reply]
Do you  have a repo or builds with active NrgModbus ?
I cannot assign a component to the GPIO. In the dropdown list for GPIO1, I don't have NrgModbus.

what should I do ?
I know Python and SQL. I do not know C... 

GPIO CONF (GPIOs ALL) 
14:29:30.806 RSL: RESULT = {"GPIOs2":{"6656":"TM1637 DIO","6688":"DLP Tx","6720":"DLP Rx","6752":"SSD1351 DC","6784":"XPT2046 CS","6816":"CSE7761 Tx","6848":"CSE7761 Rx","6880":"VL53LXX XSHUT","6912":"MAX7219 CLK","6944":"MAX7219 DIN","6976":"MAX7219 CS","7008":"TFmini+ TX","7040":"TFmini+ RX","7072":"ZC Pulse","7104":"Input","7136":"I2S DOUT","7168":"I2S BCLK","7200":"I2S WS","7232":"I2S DIN","7264":"I2S WS IN","7296":"I2S BCLK IN","7328":"Interrupt","7360":"MCP2515 CS","7392":"HRG15 Tx","7424":"HRG15 Rx","7456":"VINDRIKTNING","7488":"BL0939 Rx","7520":"BL0942 Rx","7552":"HM330X SET","7584":"Heartbeat","7616":"Heartbeat_i","7648":"74x595 SRCLK","7680":"74x595 RCLK","7712":"74x595 OE","7744":"74x595 SER","7776":"SolaxX1 RTS","7808":"Option E","7840":"SDM230 Tx","7872":"SDM230 Rx","7904":"ADC MQ","7936":"CM110x TX","7968":"CM110x RX","8000":"BL6523 Tx","8032":"BL6523 Rx","8064":"ADE7880 IRQ","8096":"Reset","8128":"MS01","8160":"SDIO CMD","8192":"SDIO CLK","8224":"SDIO D0","8256":"SDIO D1","8288":"SDIO D2","8320":"SDIO D3","8352":"Flowrate","8384":"BP5758D Clk","8416":"BP5758D Dat","8448":"SM2335 Clk","8480":"SM2335 Dat","8512":"MP3 Busy","8544":"TM1621 CS","8576":"TM1621 WR","8608":"TM1621 RD","8640":"TM1621 DAT","8672":"Relay_b","8704":"Relay_bi","8736":"I2S MCLK","8768":"ModBr Tx","8800":"ModBr Rx","8832":"ADE7953 RST","8864":"NrgModbus Tx","8896":"NrgModbus Rx"}}

**Reply by sfromis** [reply]
Seeing `"8864":"NrgModbus Tx","8896":"NrgModbus Rx"` in your output of the gpios command indicates that you have NrgModbus included in the binary you're using, those are the two gpios you need to config.

**Comment by TomaszSz7** [comment]
Thanks for your insights! 
 I've tested this change. Unfortunately, it still doesn't work.... 
UART for ESPHome is working. Maybe something is with the protocol for MODbus.
 i don't get it. 
The port settings should be the same in the driver for sdm630.

**Reply by barbudor** [reply]
It's not expected to work better but to provide more logs that can help debugging
So please provide the updated logs

What is the esp home config that you are using? 
Does that provides energy like tasmota? 

Thanks

**Reply by TomaszSz7** [reply]
`uart:
  id: mod_bus
  rx_pin: 3
  tx_pin: 1
  baud_rate: 9600
  stop_bits: 1
  
modbus:
  uart_id: mod_bus
  flow_control_pin: 5
  id: modbus1
modbus_controller:
  - id: sdm
    address: 0x0001
    modbus_id: modbus1
    update_interval: 1s
sensor:
############# SDM #############
  - platform: modbus_controller
    modbus_controller_id: sdm
    name: "Phase 1 line to neutral volts"
    id: "Phase_1_line_to_neutral_volts"
    register_type: read
    device_class: energy
    unit_of_measurement: "W"
    address: 0x0000
    value_type: FP32
    accuracy_decimals: 0
    filters:
      - multiply: 1.0
  - platform: modbus_controller
    modbus_controller_id: sdm
    name: "Phase 2 line to neutral volts"
    id: "Phase_2_line_to_neutral_volts"
    register_type: read
    device_class: energy
    unit_of_measurement: "W"
    address: 0x0002
    value_type: FP32
    accuracy_decimals: 1
    filters:
      - multiply: 1.0
  - platform: modbus_controller
    modbus_controller_id: sdm
    name: "Phase 3 line to neutral volts"
    id: "Phase_3_line_to_neutral_volts"
    register_type: read
    device_class: energy
    unit_of_measurement: "W"
    address: 0x0004
    value_type: FP32
    accuracy_decimals: 1
    filters:
      - multiply: 1.0
